admin/scripts/zvlastni/reporty/maily-dle-data-ucasti.php

```diff
@@ -0,0 +1,19 @@
+<?php
+
+require_once('sdilene-hlavicky.php');
+
+//nastavení startu pro filtr v SQL dotazu - čím vyšší číslo, tím déle účastník nebyl na GC
+$start = (int) get('start');
+isset($start) ? $start : $start = 0;
```

Comment:
tento řádek už nic nedělá

Reply:
Dáme tomu jednu obtočku, ale je to vlastně spíš dotaz na tebe:

Tento řádek obecně zajišťuje ošetření vstupu - tzn. pokud na vstupu nic není, dá tam nějakou defaultní hodnotu (nula). V budoucnu může ten report někdo použít a třeba tam zapomenout dát parametr.
Myslel jsem, že se to tak obecně dělá, ale pokud řekneš "Ne, takto se to nedělá, protože kód má být vnitřně konzistentní a tedy tyto ošetření nejsou ani vhodné, ani potřeba", tak kdo jsem já, abych to rozporoval. :-)

Reply:
Tady je ve hře víc věcí, takže to rozdělím:

1. tl;dr `get('start')` vrátí `null`, `(int)` jej přetypuje na `0`, hotovo
2. čištění vstupů je obecně nutná věc, to je správně. Standardně se to zařídí přes parametry ve volání `dbQuery($dotaz, $parametry)`, kde vyčistění zařídí samo dbQuery, ale pro LIMIT zrovna databáze parametry nepodporuje, takže se čištění musí dělat ručně (děkujeme). Funkční trik pro tento případ je přetypovat vstup na int, čímž se automaticky vyřeší, že se tam nemůže objevit nic než validní číslo.
3. defaulty jsou obecně dobrá věc, ale je to propojené s výjimkami. Default použiji, pokud nezadání věci je korektní / očekávané chování. Naopak pokud by se default použil jen při nezamýšleném chování, je lepší místo vložení defaultu generovat výjimku. Výjimka se zaloguje a můžeme zpětně řešit, proč k neočekávanému chování dochází (například máme nějakou url, kterou si určujeme v php, ale chytrý uživatel si ji vykopíruje a začne přeposílat v mail konferenci. My ji pak v php změníme, čímž ale stará verze bude vracet nesmysly, ovšem s defaultem může bez povšinutí zůstat funkční a chybná data v odkazu z mailkonference budou stále nějací lidé využívat. Bez defaultu se o tom co se děje dozvíme z výjimek a lidem se nevygeneruje nic, takže nám možná i sami napíšou, pokud budou ta data chtít).
4. osobně propaguji (bo my nejsme banka pi*) kratší a méně ošetřený kód, kde případné chyby jsou neošetřené, jen je garance, že chování je bezpečné. Co tím myslím ukážu na příkladu:

```php
$json = json_decode('http://nejaka.url');
echo 'Z nějaké url se načetlo ' . $json->jmeno;
```

vs

```php
$json = json_decode('http://nejaka.url');
if($json === false) {
  echo 'Omlouváme se, data nelze přečíst';
} else if(!isset($json->jmeno)) {
  echo 'Omlouváme se, data přišla v neočekávaném formátu. Pošlete je včetně parametru "jmeno"'.
} else {
  echo 'Z nějaké url se načetlo ' . $json->jmeno;
}
```

Předchozí kód dělá úplně to samé, ale preferuji variantu 1, protože mi přijde čitelnější a zároveň není nebezepečná (vypsání prázdného řetězce nebo nějakého nesmyslu tady nevadí). Navíc zavolání $objekt->neexistujiciAtribut vygeneruje notice, která se v GC webu přemění na výjimku a kód se zastaví (ale na to obecně nespoléhat, je potřeba věřit standardnímu chování php). K zhodnocení nebezpečnosti je potřeba vždy nějaká úvaha (XSS, SQL injection, kdo kde a jak kód bude používat, jestli nečekaná věc vygeneruje výjimku a kód skončí, nebo jen notice/warning a kód by mohl jet dál, ...).

5. ternární operátor bych používal jen pro přiřazení, tj.:

```php
// špatně
$staloSeNeco ? $var1 : $var2 = 'bla';

// správně
$var2 = $staloSeNeco ? 'ano, něco se stalo' : 'ne, nestalo se nic';

// správně, pokud chci provádět nějakou logiku
if($staloSeNeco)    vypisZeSeNecoStalo();
else                vypisZeSeNestaloNic();
```

Reply:
Díky za vysvětlení a jinak tedy fixed v poslední verzi.